I made some attempts to create a test:

Testing this with the following setup with `qemu-i386`

```sh
sudo sysctl -w vm.mmap_min_addr=0
gcc -m32 read.c -o read
qemu-i386 -g 1234 ./read
```

read.c file
```c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdint.h>
#include <errno.h>

#define PAGE_SIZE 4096
// #define MAX_ADDR ((void*)(0xFFFFFFFF - PAGE_SIZE + 1))
#define MAX_ADDR ((void*)(0xFFFFF000))

// First, allow zero-address to be mapped with
// sudo sysctl -w vm.mmap_min_addr=0
int main() {
    void* zero_addr = NULL;

    void *zero_page = mmap(zero_addr, 1, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS, -1, 0);
    if (zero_page == MAP_FAILED) {
        perror("mmap 0");
        printf("%d\n",errno);
        return 1;
    }

    printf("Zero-page pointer - %x\n", zero_page);
    *((uint32_t*)zero_page) = -1;
    printf("Read from 0: 0x%X\n", *((uint32_t*)zero_page));


    void *max_page = mmap(MAX_ADDR, 1, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS, -1, 0);
    if (max_page == MAP_FAILED) {
        perror("mmap max");
        printf("%d\n",errno);
        return 1;
    }

    printf("Max page pointer - %p\n", max_page);

    void* read_ptr = ((char *)max_page) + 0xfff;
    printf("Read pointer - %x\n", read_ptr);

    printf("One byte read: 0x%X\n", *((char*)read_ptr));
    printf("4 byte read: 0x%X\n", *((uint32_t*)read_ptr));

    return 0;
}
```

Some interesting findings:

Memory reads that read across the boundary appears to segfaults with qemu-user (even with first and last page mapped). Creating a test to check the behavior on a real CPU is a bit more difficult (due to https://stackoverflow.com/questions/47712502/why-cant-i-mmapmap-fixed-the-highest-virtual-page-in-a-32-bit-linux-process-o, this can't be done on x86_64 cpu with default linux setup).

In this case, the memory address returned in the error is the `addr` that is read
```
pwndbg> pi gdb.selected_inferior().read_memory(0xffff_ffff,2)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
gdb.MemoryError: Cannot access memory at address 0xffffffff
Error occurred in Python: Cannot access memory at address 0xffffffff
```

However, I have yet to validate this behavior with qemu-system.

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -457,7 +457,7 @@ class GDBProcess(pwndbg.dbg_mod.Process):
             if stop_addr == 0 and stop_addr < addr:
                 # We could read from the top-portion of memory, but not after wrapping around
                 # Because we are doing a partial read, read until the max address
-                return self.read_memory(addr, pwndbg.aglib.arch.ptrmask - addr)
+                return self.read_memory(addr, pwndbg.aglib.arch.ptrmask - addr + 1)
 
             if stop_addr != addr:
                 return self.read_memory(addr, stop_addr - addr)

